File: Examples/android/CMeSdkJava/app/src/main/java/me/connect/sdk/java/sample/structmessages/StructuredMessagesViewModel.java
package me.connect.sdk.java.sample.structmessages;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import java.util.List;
import java.util.concurrent.Executors;

import me.connect.sdk.java.Messages;
import me.connect.sdk.java.StructuredMessages;
import me.connect.sdk.java.message.Message;
import me.connect.sdk.java.message.MessageType;
import me.connect.sdk.java.message.StructuredMessageHolder;
import me.connect.sdk.java.sample.SingleLiveData;
import me.connect.sdk.java.sample.db.Database;
import me.connect.sdk.java.sample.db.entity.Connection;
import me.connect.sdk.java.sample.db.entity.StructuredMessage;

public class StructuredMessagesViewModel extends AndroidViewModel {
    private final Database db;
    private LiveData<List<StructuredMessage>> structMessages;

    public StructuredMessagesViewModel(@NonNull Application application) {
        super(application);
        db = Database.getInstance(application);
    }

    public LiveData<List<StructuredMessage>> getStructuredMessages() {
        if (structMessages == null) {
            structMessages = db.structuredMessageDao().getAll();
        }
        return structMessages;
    }

    public SingleLiveData<Boolean> getNewStructuredMessages() {
        SingleLiveData<Boolean> data = new SingleLiveData<>();
        checkStructMessages(data);
        return data;
    }

    private void checkStructMessages(SingleLiveData<Boolean> liveData) {
        Executors.newSingleThreadExecutor().execute(() -> {
            List<Connection> connections = db.connectionDao().getAllAsync();
            for (Connection c : connections) {
                Messages.getPendingMessages(c.serialized, MessageType.QUESTION).handle((res, throwable) -> {
                    if (res != null) {
                        for (Message msg : res) {
                            StructuredMessageHolder holder = StructuredMessages.extract(msg);
                            if (!db.structuredMessageDao().checkMessageExists(holder.getId(), c.id)) {
                                StructuredMessage sm = new StructuredMessage();
                                sm.connectionId = c.id;
                                sm.messageId = msg.getUid();
                                sm.entryId = holder.getId();
                                sm.questionText = holder.getQuestionText();
                                sm.questionDetail = holder.getQuestionDetail();
                                sm.answers = holder.getResponses();
                                sm.type = holder.getType();
                                sm.serialized = msg.getPayload();
                                db.structuredMessageDao().insertAll(sm);
                            }
                        }
                    }
                    liveData.postValue(true);
                    return null;
                });
            }
        });
    }

    public SingleLiveData<Boolean> answerMessage(int messageId, String answer) {
        SingleLiveData<Boolean> data = new SingleLiveData<>();
        answerStructMessage(messageId, answer, data);
        return data;
    }

    private void answerStructMessage(int messageId, String answer, SingleLiveData<Boolean> liveData) {
        Executors.newSingleThreadExecutor().execute(() -> {
            StructuredMessage sm = db.structuredMessageDao().getById(messageId);

            Connection con = db.connectionDao().getById(sm.connectionId);
            StructuredMessages.answer(con.serialized, sm.messageId, sm.type, sm.serialized, answer).handle((res, err) -> {
                if (err == null) {
                    sm.selectedAnswer = answer;
                    db.structuredMessageDao().update(sm);
                }
                liveData.postValue(err == null);
                return null;
            });
        });
    }


}

File: examples/android/CMeSdkJava/app_kt/src/main/kotlin/me/connect/sdk/java/samplekt/structmessages/StructuredMessagesViewModel.kt
package me.connect.sdk.java.samplekt.structmessages

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.future.await
import kotlinx.coroutines.launch
import me.connect.sdk.java.Messages
import me.connect.sdk.java.StructuredMessages
import me.connect.sdk.java.message.MessageType
import me.connect.sdk.java.samplekt.SingleLiveData
import me.connect.sdk.java.samplekt.db.Database
import me.connect.sdk.java.samplekt.db.entity.StructuredMessage
import me.connect.sdk.java.samplekt.wrap


class StructuredMessagesViewModel(application: Application) : AndroidViewModel(application) {
    private val db = Database.getInstance(application)
    private val structMessagesLiveData by lazy {
        db.structuredMessageDao().getAll()
    }

    fun getStructuredMessages(): LiveData<List<StructuredMessage>> = structMessagesLiveData

    fun getNewStructuredMessages(): SingleLiveData<Boolean> {
        val data = SingleLiveData<Boolean>()
        checkStructMessages(data)
        return data
    }

    private fun checkStructMessages(liveData: SingleLiveData<Boolean>) = viewModelScope.launch(Dispatchers.IO) {
        try {
            val res = Messages.getPendingMessages(MessageType.QUESTION).wrap().await()
            res.forEach { msg ->
                val holder = StructuredMessages.extract(msg)
                val pwDid: String = msg.pwDid
                if (!db.structuredMessageDao().checkMessageExists(pwDid)) {
                    val sm = StructuredMessage(
                            pwDid = pwDid,
                            messageId = msg.uid,
                            entryId = holder.id,
                            questionText = holder.questionText,
                            questionDetail = holder.questionDetail,
                            answers = holder.responses,
                            serialized = msg.payload,
                            type = holder.type
                    )
                    db.structuredMessageDao().insertAll(sm)
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            liveData.postValue(true)
        }
    }


    fun answerMessage(messageId: Int, answer: String): SingleLiveData<Boolean> {
        val data = SingleLiveData<Boolean>()
        answerStructMessage(messageId, answer, data)
        return data
    }

    private fun answerStructMessage(messageId: Int, answer: String, liveData: SingleLiveData<Boolean>) = viewModelScope.launch(Dispatchers.IO) {
        try {
            val sm = db.structuredMessageDao().getById(messageId)
            val con = db.connectionDao().getByPwDid(sm.pwDid)
            StructuredMessages.answer(con.serialized, sm.messageId, sm.type, sm.serialized, answer).wrap().await()
            sm.selectedAnswer = answer;
            db.structuredMessageDao().update(sm)
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            liveData.postValue(true)
        }
    }
}